src/System.Net.WebSockets.WebSocketProtocol/src/System/Net/WebSockets/ManagedWebSocketExtensions.cs

```diff
@@ -49,6 +49,32 @@ internal static Task<int> ReadAsync(this Stream stream, Memory<byte> destination
                 }
             }
         }
+
+        internal static ValueTask WriteAsync(this Stream stream, ReadOnlyMemory<byte> source, CancellationToken cancellationToken = default)
```

Comment:
These extension methods seem really weird to me, but I guess they already exist (at least for Read), so that's okay, I suppose?  What exactly is going on here?

Reply:
> What exactly is going on here?

We have the ManagedWebSocket.cs implementation that's compiled into multiple assemblies.  It's compiled into System.Net.WebSockets.dll, which is part of netcoreapp and can always use all the latest bells and whistles, including the new Read/WriteAsync overloads on Stream.  It's also copiled into System.Net.WebSockets.WebSocketProtocol.dll, which is an OOB assembly we release specifically to target netstandard2.0, and in particular so that ASP.NET can use ManagedWebSocket whether targeting netstandard2.0 or netcoreapp.  When built in this configuration, ManagedWebSocket can't use the new Stream overloads, so we have two choices: either use ifdef'ing in ManagedWebSocket to have two different implementations for the parts of the code where we use the new functionality, or use partial classes to provide polyfills when building for netstandard.  We chose the latter, and that's what these are, implementations of the new APIs, with the same signature, but that aren't actually overloads and are instead just static extensions with the same signature otherwise.  The implementation is essentially the same as what the base Stream implementation provides in netcoreapp: try to extract the array if there is one and then just delegate to the array overload, or fall back to renting an array, copying data, and using the array-based overload.

Reply:
Thanks, that makes sense.